Added throttling to token generation (Closes #48)

demestav opened the pull request:
As the #48 issue describes, it is beneficial to optionally throttle the token generation, to prevent accidental or malicious abuse of the token delivery medium.

The current state of this pull request is functional but not fully tested. If the maintainer sees some value to this, I could use some feedback to finalize this.

psagers commented on src/django_otp/models.py:
```diff
@@ -226,10 +225,28 @@ class SideChannelDevice(Device):
         default=timezone.now,
         help_text="The timestamp of the moment of expiry of the saved token.",
     )
+    last_generated = models.DateTimeField(
```
This will trigger a migration in existing devices, which may require a semantic version bump. Is it possible for it to live in `GenerationThrottlingMixin`? Also, should it get reset on a successful validation?

demestav replied:
> This will trigger a migration in existing devices, which may require a semantic version bump

I believe it belongs together with the rest of the token attributes (`token` and `valid_until`). Indeed it will trigger a migration, but since it is `null=True`, I don't think it will break anything on existing installations. Therefore, it would not need a major version bump, unless of course that is your policy.

> Is it possible for it to live in GenerationThrottlingMixin?

In my opinion (for the reasons listed in my first comment), this mixin should be the default for the interactive devices. That includes the EmailDevice that ships with the package. Therefore, that would trigger a migration anyway.

> Also, should it get reset on a successful validation?
Indeed, I will add this.

On a final note, I was thinking how we can do this without any migration and one approach is to use the `valid_until`. Which means, the user can generate another token, only if the previous token has expired. However, I decided that it is better to have a different expiration for generation. But I mentioning it here, in case you prefer to go this way.

psagers replied:
Suppose django-otp adds a field to SideChannelDevice, but some other project similar to django-otp-twilio doesn't get updated for a little while? What happens when an unsuspecting project gets a new field on a third-party model that they have no direct control over?

demestav replied:
> Suppose django-otp adds a field to SideChannelDevice, but some other project similar to django-otp-twilio doesn't get updated for a little while? What happens when an unsuspecting project gets a new unmigrated field on a third-party model that they have no direct control over?

Sorry if this is trivial, but I am trying to follow this through; the user installs django-otp which includes the new field in SideChannelDevice. Also installs a project which is unaware of this change e.g. "django-otp-twillio". Since it uses `SideChannelDevice` model from django-otp, this will trigger a migration for the `TwilioSMSDevice`. The migration can be applied without having any effect on the application. What am I missing?

psagers replied:
Let's think through what actually happens here. You have a project with a dependency on django-otp-twilio (in this example). You update django-otp and run makemigrations. Django sees that TwilioSMSDevice has acquired a new field and needs a migration. Where is it going to write this migration file? In the installed django-otp-twilio package in your development virtualenv? How does that help us?

Even if you could somehow generate a migration for a third-party app and check it in somewhere, what happens when django-otp-twilio updates itself and adds the official migration with its own name? You update again and migrate and Django tries to apply a new migration to add a field that already exists in your database table. Now I guess you have to `migrate --fake` or something. Total chaos.

I haven't tried simulating this, but I assume that's what would happen. The burden of proof is on demonstrating conclusively that this can be made to work seamlessly and transparently.

demestav replied:
> Even if you could somehow generate a migration for a third-party app and check it in somewhere, what happens when django-otp-twilio updates itself and adds the official migration with its own name? You update again and migrate and Django tries to apply a new migration to add a field that already exists in your database table. Now I guess you have to `migrate --fake` or something. Total chaos.

Indeed that is an undesirable case and should be avoided. 

1. I am curious, under the same terms, does that mean that none of the existing models can be altered, ever? 
2. Does this mean that the whole PR falls apart? Because even if the migration is triggered by the mixin, the same case applies, if the EmailDevice uses the mixin by default.

demestav replied:
> Is it possible for it to live in GenerationThrottlingMixin?

Completed

> Also, should it get reset on a successful validation?

Completed.

psagers commented on src/django_otp/models.py:
```diff
@@ -226,10 +225,28 @@ class SideChannelDevice(Device):
     class Meta:
         abstract = True
 
+    def generate_is_allowed(self):
```
I think this stub could live in the root Device model, alongside verify_is_allowed. This would also simplify OTPAuthenticationFormMixin.

demestav replied:
It would be more tidy to have it in the same place as `verify_is_allowed`. However,  I thought to place it in `SideChannelDevice` because `generation_is_allowed` only applies only to the devices that generate token.

demestav replied:
I am not expert on the subject, I don't know what other channels are usually used to deliver a token, apart from email or SMS. But from my point of view, both email and sms should be throttled. But you are the expert on the subject so let me know if this is not the case and how to move this forward.

psagers replied:
> I don't know what other channels are usually used to deliver a token, apart from email or SMS.

This is a good logical fallacy to become familiar with: it's sometimes called proof by lack of imagination.

This is a framework. There are an infinite number of devices that might be written and the only rule is that they must inherit from `Device`. If you're thinking just in terms of the specific devices you know about or imagine, then you're thinking too small. The question is what's the right thing to do for every device that has ever or will ever be written.

Mixins and things like `SideChannelDevice` are convenience classes that help people write standard categories of devices in standard ways. But we can never imagine all possibilities, so any device authors are free to ignore the helpers and start from scratch. That's why `generate_token` and `verify_is_allowed` are on the root `Device` class: they don't apply to *every* device, but they could apply to *any* device.

demestav replied:
I considered `SideChannelDevice` as the class to subclass in case someone needed a device that uses token generation, verification and expiration.



> Mixins and things like `SideChannelDevice` are convenience classes that help people write standard categories of devices in standard ways. But we can never imagine all possibilities, so any device authors are free to ignore the helpers and start from scratch. That's why `generate_token` and `verify_is_allowed` are on the root `Device` class: they don't apply to _every_ device, but they could apply to _any_ device.

I see your point and agree. As you suggested, the `generate_is_allowed` should be in the `Device`.

psagers commented on src/django_otp/forms.py:
```diff
@@ -134,6 +139,19 @@ def _chosen_device(self, user):
+        try:
+            generate_is_allowed, extra = device.generate_is_allowed()
+        except AttributeError:
+            raise forms.ValidationError(
```
This error seems like a mismatch. But it can just be removed if `generate_is_allowed` is moved to the root class.

demestav replied:
As noted above, `generate_is_allowed` applies only to devices that generate token (`SideChannelDevice`). Therefore, if somehow the `generate_is_allowed` is called on a device that does not have it we notify the user.

psagers commented on src/django_otp/models.py:
```diff
@@ -290,6 +309,35 @@ class VerifyNotAllowed:
+        if (
+            not self.last_generated
+            or (dt_now - self.last_generated).total_seconds()
+            > settings.OTP_GENERATION_INTERVAL
```
`ThrottlingMixin` has a `get_throttle_factor` method that subclasses must implement. Each device is responsible for managing its own setting. Default values are conventionally defined in relevant conf.py modules.

demestav replied:
Sounds good. I will add `get_generation_interval` method.

demestav replied:
> I will add `get_generation_interval` method.

Completed

psagers reviewed (changes requested):
Looks like a good start.

If the interaction between `GenerationThrottlingMixin` and `SideChannelDevice` gets complicated, it might worth considering whether [singledispatch](https://docs.python.org/3/library/functools.html#functools.singledispatch) could be used to cut through some of the MRO subtleties.

demestav commented:
Thank you for the quick response and feedback. I will respond to each comment in their respective dialog.

> If the interaction between `GenerationThrottlingMixin` and `SideChannelDevice` gets complicated, it might worth considering whether [singledispatch](https://docs.python.org/3/library/functools.html#functools.singledispatch) could be used to cut through some of the MRO subtleties.

Let's see how it turns out with the Mixin reconsider.

psagers commented on src/django_otp/models.py:
```diff
@@ -290,6 +303,68 @@ class VerifyNotAllowed:
+        Update the timestamp of the last token generation to the current time.
+        """
+        self.last_generated_timestamp = timezone.now()
+        super().generate_token(*args, **kwargs)
```
Is `generate_token` guaranteed to exist on all devices? This appears to be a secret dependency on `SideChannelDevice`.

demestav replied:
This should be resolved now.

psagers commented:
I'll take another look at this once it includes documentation and test coverage.

demestav commented:
> I'll take another look at this once it includes documentation and test coverage.

I will re-request a review once I have something ready.

demestav commented:
Had to rebase to resolve conflicts with `master`. Re-requesting review.

psagers commented on src/django_otp/models.py:
```diff
@@ -293,6 +306,68 @@ class VerifyNotAllowed:
+        else:
+            return False, {
+                'reason': "COOLDOWN_DURATION_PENDING",
+                'next_generation_at': dt_now
```
Should `dt_now` be `self.last_generated_timestamp` here?

demestav replied:
Indeed! I have corrected it and added a test for the message as well.

psagers commented on src/django_otp/tests.py:
```diff
@@ -164,6 +165,45 @@ def test_verify_is_allowed(self):
             self.assertEqual(data3, None)
 
 
+class GenerationThrottlingTestMixin:
```
It looks like this is based on `ThrottlingTestMixin`, which makes sense. However, it's asserting on `len(mail.outbox)`, which is email-specific. This will presumably need at least one abstract method to actually be generic.

demestav replied:
I agree. I have added generic tests and moved email specific tests to the email test file ([29ebe3e](https://github.com/django-otp/django-otp/pull/122/commits/29ebe3e64a5daca52a4272edb0cc4ef02df8526c)).

psagers commented on src/django_otp/tests.py:
```diff
@@ -164,6 +165,45 @@ def test_verify_is_allowed(self):
+            # Assert that no email is sent
+            self.assertEqual(len(mail.outbox), 1)
+
+        with freeze_time() as frozen_time:
```
This test should work in practice as-is, but would it make sense to use this context for the entire method body? The point is to control time for the whole sequence of operations.

demestav replied:
Yes, it makes perfect sense. I have corrected this in all tests ([29ebe3e](https://github.com/django-otp/django-otp/pull/122/commits/29ebe3e64a5daca52a4272edb0cc4ef02df8526c)).

psagers commented on src/django_otp/plugins/otp_email/models.py:
```diff
@@ -57,34 +63,50 @@ def generate_challenge(self, extra_context=None):
-                context
+        generate_allowed, data_dict = self.generate_is_allowed()
+        if generate_allowed:
+            self.generate_token(valid_secs=settings.OTP_EMAIL_TOKEN_VALIDITY)
```
This method is now doing two big things. Perhaps it's time to move the body of `if generate_allowed:` into a private method for clarity. See https://github.com/django-otp/django-otp-twilio/blob/9e1aab519c926a3d281292d549ce2040c821fe1a/src/otp_twilio/models.py#L55, e.g.

demestav replied:
Created a private method for that part ([dbe61d7](https://github.com/django-otp/django-otp/pull/122/commits/dbe61d7fe664ad0fc7ebeb74b630ea682cc970b6)).

psagers reviewed (changes requested):
Structurally, this is looking pretty solid. I added a few suggestions for fixes and cleanup. And of course there's still the documentation.

Remember to run `hatch run check` (and, if necessary, `hatch run fix`) before checking in.

demestav commented:
> Remember to run `hatch run check` (and, if necessary, `hatch run fix`) before checking in.

Yes, sorry about that, will do.

> I added a few suggestions for fixes and cleanup. 

Thank you for the feedback. I believe I have addressed all open issues and commented on relevant conversations above.

> And of course there's still the documentation.

I used `ThrottlingMixin` as a guide and replicated for this Mixin. I was not sure whether `generate_is_allowed` should be listed in "members" in the "Helpers" sections.

Please note 
- I have renamed the feature as Cooldown to avoid reusing the word "Throttling" which is used in the `ThrottlingMixin` and to shorten the name a bit. I believe `CooldownMixin` is a more appropriate name for this feature.
- I have set the default value for email cooldown to 60 seconds.

psagers commented on src/django_otp/plugins/otp_email/models.py:
```diff
@@ -57,7 +59,27 @@ def generate_challenge(self, extra_context=None):
+
+    def _deliver_token(self, extra_context):
         self.generate_token(valid_secs=settings.OTP_EMAIL_TOKEN_VALIDITY)
+        self.last_generated_timestamp = timezone.now()
```
This never gets saved. As importantly, the tests aren't catching this (see companion note in the tests). 

(Hint: try swapping these two lines. Adding the technically redundant `commit=True` will help to highlight the fact that you're relying on `generate_token` to save the model instance for you.)

psagers commented on src/django_otp/tests.py:
```diff
@@ -164,6 +164,78 @@ def test_verify_is_allowed(self):
+        """
+        with freeze_time():
+            self.device.generate_challenge()
+            allowed, details = self.device.generate_is_allowed()
```
If you insert `self.device.refresh_from_db()` between these lines, the test fails (`last_generated_timestamp` never gets committed).

demestav commented:
> There's a critical bug in the email integration

Indeed, I missed this. I have added `cooldown_set` method to the mixin to keep it consistent, and updated the email integration code.

> liberal use of `self.device.refresh_from_db()` or equivalent in the tests will allow them to catch these sorts of bugs.

I have updated the tests.

psagers reviewed (approved):
Looks like it's ready. Thanks for all the work.

demestav commented:
> Looks like it's ready. Thanks for all the work.

I am glad we got this through! Thank you for your valuable feedback and patience.

@psagers I know this is merged now, but it occurred to me that maybe the default value for `OTP_EMAIL_COOLDOWN_DURATION` should have been 0 in order to disable the cooldown feature by default? The reason I am suggesting this, is that people that are already using the email plugin, and upgrade, and they are unware of this feature, they will unintentionally activate cooldown which might affect how their application works.

psagers commented:
I considered that, although `OTP_EMAIL_THROTTLE_FACTOR` also has a default value, so it's consistent. 60 seconds isn't a very aggressive cooldown and I noted the new behavior in the changelog, so erring on the side of safety seems reasonable here.
